### PR TITLE
chaindb: migrate and fix total supply calculation. see #394.

### PR DESCRIPTION
**Pull request by chjj**

### CODE REVIEW

File: lib/blockchain/layout.js [hunk 1]
```diff
@@ -29,6 +29,7 @@ const bdb = require('bdb');
  *   w[height] -> name undo
  *   s -> tree state
  *   f -> bit field
+ *   M[migration-id] -> migration
```
Comment (pinheadmz) [inline lib/blockchain/layout.js]: nit: should specify `dummy` ?

**Comment by pinheadmz** [comment]
@kilpatty are you able to review this as well? Maybe run it with hnscan indexer?

**Comment by kilpatty** [comment]
LGTM on mobile right now, but I’ll run a build of hnscan-backend with this in an hour or two and report back.

**Comment by pinheadmz** [comment]
Ran both master and branch with this extra log:

```diff
diff --git a/lib/blockchain/chaindb.js b/lib/blockchain/chaindb.js
index 214b91bd..7f3198ce 100644
--- a/lib/blockchain/chaindb.js
+++ b/lib/blockchain/chaindb.js
@@ -1486,6 +1486,8 @@ class ChainDB {
     // Connect block and save data.
     await this.saveBlock(entry, block, view);
 
+    console.log(`height: ${entry.height} current: ${this.state.value} new: ${this.pending.value}    diff: ${this.pending.value - this.state.value}`)
+
     // Commit new chain state.
     this.put(layout.R.encode(), this.pending.commit(hash));
   }
```

Confirmed on master there were 218 blocks with negative value applied to chainstate and on branch, only 40.

I'm still kind of surprised there were any at all... what is destroying these coins?

**Comment by chjj** [comment]
@pinheadmz, try this:

``` diff
diff --git a/lib/blockchain/chaindb.js b/lib/blockchain/chaindb.js
index 9e2c9af7..e1248742 100644
--- a/lib/blockchain/chaindb.js
+++ b/lib/blockchain/chaindb.js
@@ -1577,6 +1577,16 @@ class ChainDB {
     // Connect block and save data.
     await this.saveBlock(entry, block, view);

+    const added = this.pending.value - this.state.value;
+
+    if (added < 0) {
+      const burned = this.pending.burned - this.state.burned;
+
+      this.logger.info('height=%d current=%d new=%d added=%d burned=%d',
+                       entry.height, this.state.value, this.pending.value,
+                       added, burned);
+    }
+
     // Commit new chain state.
     this.put(layout.R.encode(), this.pending.commit(hash));
   }
```

In the negative blocks, you can clearly see that there are burned coins (REGISTERs).

Example:

```
[info] (chaindb) height=5773 current=266868346659591 new=266868345659591 added=-1000000 burned=2001000000
```

**Comment by pinheadmz** [comment]
@chjj is there any chance a node could get interrupted during migration and be stuck in a weird state? i.e. Is the commitment to the new chainstate atomic along with the migration ID marker? (looks like it is a la `batch` just wanna confirm)

**Comment by chjj** [comment]
@pinheadmz, it should be completely atomic.

**Pull request merged**